Gael OCL style conventions edits read run orbit node (#1098)
OffchainLabs/arbitrum-docs: arbitrum-docs/node-running/how-tos/running-an-orbit-node.mdx

---
title: 'How to run a full node for an Orbit chain'
description: 'Learn how to run an Arbitrum orbit node on your local machine'
author: Mehdi Salehi
sme: Mehdi Salehi
content_type: how-to
---

This how-to provides step-by-step instructions for running an Orbit node on your local machine.

> ### Minimum Hardware Configuration

Reviewer: Shouldn't titles be sentence-case? (or did this change?)
i.e. `Minimum hardware configuration`

Author: You're right. That's fixed.

- **RAM**: 8-16 GB
- **CPU**: 2-4 core CPU (For AWS: t3 xLarge)
- **Storage**: This depends on the Orbit chain and its traffic overtime.

## Prerequisites

- Latest Docker Image: <code>@latestNitroNodeImage@</code>

## Required Parameters

### 1. Parent Chain Parameters

The `parent-chain` argument needs to provide a standard RPC endpoint for an EVM node, whether self-hosted or obtained from a node service provider:

```shell
--parent-chain.connection.url=<Parent chain RPC URL>
```
The RPC endpoint must direct to <a data-quicklook-from="arbitrum-one">Arbitrum One</a>, <a data-quicklook-from="arbitrum-nova">Arbitrum Nova</a>, or<a data-quicklook-from="arbitrum-sepolia">Arbitrum Sepolia</a>.

:::note

[Public Arbitrum RPC endpoints](/build-decentralized-apps/reference/01-node-providers.md#arbitrum-public-rpc-endpoints) rate-limit connections. To avoid hitting a bottleneck, you can  run a local node for the parent chain or rely on third-party RPC providers.

::: 

### 2. Child Chain Parameters 

> In the Arbitrum Orbit context, the child chain is an L2 or an L3 Orbit chain, and the required parameters are `chain-info` and `chain-name`

Reviewer: ```suggestion
In the Arbitrum Orbit context, the child chain is an L2 or an L3 Orbit chain, and the required parameters are `chain.info-json` and `chain.name`
```

> #### 1. `chain.info`

Reviewer: ```suggestion
#### 1. `chain.info-json`
```

`--chain.info-json` is a JSON string that contains required information about the Orbit chain. 

```shell
--chain.info-json=<Orbit Chain's chain info>
```

> An example of `chain-info` is available in the next section.

Reviewer: ```suggestion
An example of `chain.info-json` is available in the next section.
```

An example of `chain-info` is available in the next section.

  ```shell
  docker stop --time=300 $(docker ps -aq)
  ```
#### 2. `chain.name`

`--chain.name` is a mandatory flag that needs to match the chain name used in `--chain.info-json`:

```shell
--chain.name=<My Arbitrum L3 Chain> 
```

#### 3. AnyTrust Chains

For <a data-quicklook-from="arbitrum-anytrust-chain">Anytrust chains</a>, you need to to add the following flags to the command or configuration:

```shell 
--node.data-availability.enable
```

And: 

## Note on permissions
```shell 
--node.data-availability.rest-aggregator.urls=<A list of DAS REST endpoints>
```

The Docker image is configured to run as non-root UID 1000. This means if you are running in Linux or OSX and you are getting permission errors when trying to run the docker image, run this command to allow all users to update the persistent folders
Or: 

```shell
mkdir /data/arbitrum
chmod -fR 777 /data/arbitrum
--node.data-availability.rest-aggregator.online-url-list=<A url that returns a list of the DAS REST endpoints>
```

> ## Optional parameters

Reviewer: This section is now missing, can you bring it back at the end of the page?

Author: Yes, it's done.

> ### 3.Important Ports

Reviewer: I like the table structure, although its current info might be a bit confusing for beginners. How about this?
- Change rows to: RPC (http) | RPC (ws) | Sequencer feed (notice I changed the websocket one to one place above and changed the label)
- Only have the Protocol and Port columns
- Add an asterisk (or something similar to `RPC (ws)` with a note below the table to add the needed ports to enable the websocket.

- **RAM**: 8-16 GB
- **CPU**: 2-4 core CPU (For AWS: t3 xLarge)
- **Storage**: This depends on the Orbit chain and its traffic overtime.

### Prerequisites

- Latest Docker Image: <code>@latestNitroNodeImage@</code>

> ## Required parameters

Reviewer: nitpick: Adding the comment here, although it's more for previous headers.
We are using level 3 headers after the level 1 title, and then here we start using level 2 headers.
We should either change the previous header to level 2, or create a level 2 header before those.

Author: That's a good point.
TBH I don't always use MD headers incrementally since they are just templates for what needs to be outlined. In this article, prerequisites are not the "root" of the structure, but they are among the first elements that need to be communicated. Does it make sense?

Reviewer: Interesting 🤔 
IMO, headers should outline the structure of the information of a page (with a specific order). If there's a subsection that doesn't seem to suit the specific header it should have, there might be something missing in that subsection (an introduction, or some content before that subsection?). Happy to discuss further though, I find your point of view interesting (hadn't thought about it).

In this specific case, I see your point in that the current content in prerequisites might not fit the place it currently lives in, and the page structure overall could be improved. Let me propose something: what about having a first section titled "Prerequisites" with the minimum hardware requirements, and the Nitro version needed? (no need for extra subtitles there). That prerequisites would feel more like an `h2` (at least from my point of view). What do you think?

Adding another note on headers, when moving MD headers to HTML, keeping them ordered also improves SEO in general as it helps search engines parse the information of the page correctly.

Author: I was unaware of the SEO implications of not using an incremental Header order; thanks for pointing this out.
I have no opinion on the best way to display a prerequisite paragraph. I just don't want it to steal the reader's attention from the main paragraphs. 
What header template would you propose for the `prerequisite` paragraph?

Reviewer: Yeap, better. Thanks! 🙏

### 1. Parent chain parameters

The `parent-chain` argument needs to provide a standard RPC endpoint for an EVM node, whether self-hosted or obtained from a node service provider:

```bash
--parent-chain.connection.url=<Parent chain RPC URL>
```

> The RPC endpoint must direct to <a data-quicklook-from="arbitrum-one">Arbitrum One</a>, <a data-quicklook-from="arbitrum-nova">Arbitrum Nova</a>, or<a data-quicklook-from="arbitrum-sepolia">Arbitrum Sepolia</a>.

Reviewer: ```suggestion
The RPC endpoint must direct to the chain that serves as the settlement layer of the Orbit chain.
```
This has changed a bit. Orbit chains can now settle to any EVM-based chain (Ethereum, Arb chains, Base, OP, ...).
I'm adding a suggestion, but feel free to change it in whatever way you prefer.

Author: Merged

Author: Sorry about that!

:::note

[Arbitrum RPC endpoints](/node-running/node-providers.mdx/#rpc-endpoints) rate-limit connections. To avoid hitting a bottleneck, you can  run a local node for the parent chain or rely on third-party RPC providers.

::: 

### 2. Child chain parameters 

> In the Arbitrum Orbit context, the child chain is an L3 Orbit chain, and the required parameters are `chain-info` and `chain-name`

Reviewer: This is not always the case. An Orbit chain can be an L2.

Author: Merged.

#### 1. `chain.info`

`--chain.info-json` is a JSON string that contains required information about the Orbit chain. 

```bash
--chain.info-json=<Orbit Chain's chain info>
```

An example of `chain-info` is available in the next section.

#### 2. `chain.name`

`--chain.name` is a mandatory flag that needs to match the chain name used in `--chain.info-json`:

```bash
--chain.name=<My Arbitrum L3 Chain> 
```

> For an <a data-quicklook-from="arbitrum-anytrust-chain">Anytrust chain</a>, it is essential to provide `--node.data-availability.enable`.

Reviewer: Can we add a subsection for the AnyTrust chains? They need more than just that parameter. You can see what else is needed in the "Examples" below.

Author: I tried to made a subsection out of it, wdyt?

Reviewer: Perfect, thanks!

### 3.Important ports

- RPC: `8547`
- Sequencer Feed: `9642`
- WebSocket: `8548`
  - WS port `8548` needs extra args to be opened. Please use these flags:
    - `--ws.port=8548`
    - `--ws.addr=0.0.0.0`
    - `--ws.origins=\*`

### 4. Putting it all together

- When running a Docker image, an external volume should be mounted to persist the database across restarts. The mount point inside the Docker image should be `/home/user/.arbitrum`

- Example:

  ```bash wordWrap=true

>   Docker run --rm -it  -v /some/local/dir/arbitrum:/home/user/.arbitrum -p 0.0.0.0:8547:8547 -p 0.0.0.0:8548:8548 @latestNitroNodeImage@ --parent-chain.connection.url=<Parent chain RPC URL> --chain.id=<L3ChainId> --chain.name=<My Arbitrum L3 Chain> --http.api=net,web3,eth --http.corsdomain=* --http.addr=0.0.0.0 --http.vhosts=* --chain.info-json=<Orbit Chain's chain info>

Reviewer: ```suggestion
  docker run --rm -it  -v /some/local/dir/arbitrum:/home/user/.arbitrum -p 0.0.0.0:8547:8547 -p 0.0.0.0:8548:8548 @latestNitroNodeImage@ --parent-chain.connection.url=<Parent chain RPC URL> --chain.id=<OrbitChainId> --chain.name=<My Arbitrum Orbit Chain> --http.api=net,web3,eth --http.corsdomain=* --http.addr=0.0.0.0 --http.vhosts=* --chain.info-json=<Orbit Chain's chain info>
```

`docker` should be lowercase, and removing L3 references

  ```

>   - Please note: for Anytrust chains, it is essential to add the following flags to the command or configuration:

Reviewer: This could be moved to the section above, since it's not actually an example

Author: Could you specify what isn't an example?

Reviewer: I don't remember how this looked like before (there was an explanation of the Anytrust parameters in the "examples" section), but it's now good 💪

    ```bash 
    --node.data-availability.enable
    ```

    And: 

    ```bash 
    --node.data-availability.rest-aggregator.urls=<A list of DAS REST endpoints>
    ```
    
    Or: 
    
    ```bash
    --node.data-availability.rest-aggregator.online-url-list=<A url that returns a list of the DAS REST endpoints>
    ```

- When shutting down the Docker image, it is important to allow a graceful shutdown so that the current state can be saved to disk. Here is an example of how to do a graceful shutdown of all Docker images currently running

  ```bash

>   Docker stop --time=300 $(Docker ps -aq)

Reviewer: ```suggestion
  docker stop --time=300 $(docker ps -aq)
```

  ```

### Note on permissions

- The Docker image is configured to run as non-root `UID 1000`. If you are running Linux or macOS and you are getting permission errors when trying to run the Docker image, run this command to allow all users to update the persistent folders:

>   ```bash

Author: It's indeed more generic, I think. And the resulting highlights are the same.

>   ```bash

Author: Merged.
